### DISCUSSION TITLE
Load mdx files from another folder and render as pages

### DISCUSSION THREAD

**Discussion by openscript** [vikejs/vike#1052] [Discussion]
Is it somehow possible to load mdx pages from another path (outside of the vite-plugin-ssr domain)? Should I just symbolically link the other folder into the file system routing?

**Comment by brillout** [comment]
I'm not 100% sure but I think you actually can import files outside of Vite's `config.root`. Have you tried?

Symlinking inside a monorepo also seems fine.

**Reply by openscript** [reply]
Does https://github.com/brillout/vite-plugin-ssr/discussions/1052#discussioncomment-6646402 show the idea you had in mind to import files outside of Vite's `config.root`?

**Reply by brillout** [reply]
https://github.com/brillout/vite-plugin-ssr/discussions/1052#discussioncomment-6649593

**Comment by monteeya** [comment]
I have a related question... I have a directory of markdown files (i.e. slug.md), is it possible to use them directly as VPS server pages (without renaming/symlinking them to something like slug.page.server.md)?

**Reply by brillout** [reply]
Not currenlty but it's on the radar. Feel free to create a new feature request for it. Also, in case [sponsoring](https://github.com/sponsors/brillout) is an option for your company, then I'd implement this within a week.

**Reply by monteeya** [reply]
It's for a personal project, I'm evaluating VPS as an option.

Maybe exposing some kind of hook that would give us control over loaded pages would be enough, what do you think?

**Reply by brillout** [reply]
I created a new feature request for it: https://github.com/brillout/vite-plugin-ssr/issues/1056.

**Reply by monteeya** [reply]
Thank you, I'll try to get more familiar with the project, but just to add a bit of background... I've been coding for more than three decades now, and over time, I've grown to hate opinionated frameworks. I was searching for something unopinionated on top of Vite, and found VPS, which is in terms of philosophy exactly what I want, but I have to check if the architecture is flexible enough not to get in my way (having full control over how file system routing works is crucial).

**Reply by brillout** [reply]
> having full control over how file system routing works is crucial

Do you mind elaborating?

> I've grown to hate opinionated frameworks.

I share the sentiment... frameworks tend to become tentacular and overreaching.

**Reply by monteeya** [reply]
> Do you mind elaborating?

I meant being able to implement custom routing logic (e.g. Metalsmith is awesome in that regard since it gives you control over the whole pipeline, but things like HMR make this more difficult in VPS - I have to get more familiar with the codebase to be able to give any constructive feedback in that regard).

**Reply by brillout** [reply]
> Metalsmith

This looks quite neat and flexible indeed.

>  I have to get more familiar with the codebase to be able to give any constructive feedback in that regard

👍 If you hit limitations, then feel free to elaborate on concrete use cases where you need more flexibilty than vite-plugin-ssr offers. Looking forward to it.

**Comment by openscript** [comment]
What also works for me is to load the files programmatically like follow `index.page.server.ts`:

```typescript
import { renderToString } from "react-dom/server";

export async function onBeforeRender() {
  const markdown = await import("../../../../../assets/content/de-CH/guides/index.mdx");
  return {
    pageContext: {
      pageProps: {
        markdown: renderToString(markdown.default({ components: { myCustomComponents } })),
      },
    },
  };
}
```

and inside the `index.page.tsx`:

```tsx
type PageProps = {
  markdown: string;
};

export function Page({ markdown }: PageProps) {
  return (
    <>
      <section dangerouslySetInnerHTML={{ __html: markdown }}></section>
    </>
  );
}
```

Is this advisable?

**Reply by brillout** [reply]
How about these?
 - Symlinking `../../../../../assets/content/de-CH/guides/index.mdx` to `/pages/guides/index/+Page.mdx`.
 - Re-exporting:
   ```js
    // /pages/guides/index/+Page.js
    export { default } from '../../../../../assets/content/de-CH/guides/index.mdx'
   ```

**Reply by brillout** [reply]
Just seeing you aren't using the [V1 design](https://vite-plugin-ssr.com/migration/v1-design), but you can do the same with the 0.4 design.

**Reply by openscript** [reply]
Thank you for your answer. I have multiple pages (I started with one, so I'm not overwhelmed) and also the folder structure I want to use. I currently have this `index.page.server.ts`:

```ts

import { Title } from "@timelet/ui";
import React from "react";
import { renderToString } from "react-dom/server";
import { MDXModule } from "mdx/types";
import { PageContext } from "../../../renderer/types";

const pages = import.meta.glob<MDXModule>("../../../../../assets/content/de-CH/**/*.mdx");

function createPageContext(page: MDXModule) {
  return {
    pageContext: {
      pageProps: {
        markdown: renderToString(page.default({ components: { Title: (props) => React.createElement(Title, props) } })),
      },
    },
  };
}

export async function prerender() {
  const renderPages = Object.keys(pages).map(async (p) => {
    const page = await pages[p]();
    const url = p.replace("../../../../../assets/content/de-CH", "").replace(".mdx", "").replace("index", "");
    return {
      url,
      ...createPageContext(page),
    };
  });

  return await Promise.all(renderPages);
}

export async function onBeforeRender(pageContext: PageContext) {
  const path = pageContext.urlPathname.match(/^\/docs\/?$/) ? "/docs/index" : pageContext.urlPathname;
  const match = pages[`../../../../../assets/content/de-CH${path}.mdx`];
  const page = await match();
  return createPageContext(page);
}

```

I haven't used vite-plugin-ssr before and I'm evaluating to replace some of our Gatsby apps. I'm happy how it's going and I appreciated the descriptive error messages of vite-plugin-ssr more than once. 

My questions are:

1.  I feel like I have quite a lot of boilerplate code for this usecase. Probably I can provide my custom components via `<MDXProvider>` to the mdx pages?
2. How to work with the approach you mentioned above with multiple pages?

**Reply by brillout** [reply]
AFAICT the easiest would be to symlink the whole directory that contains the `.mdx` files and then use https://github.com/brillout/vite-plugin-ssr/issues/1056.

In the meantime you can symlink each file, or keep the solution you already have.

> I feel like I have quite a lot of boilerplate code for this usecase.

True. Although, as long as you don't use internals APIs I think it's fine. One idea of vite-plugin-ssr is to provide lower-level interfaces so that you can build custom solutions, which is what you're doing. So, in a way, what you're doing is actually idiomatic.

> ```js
>       ...createPageContext(page),
> ```

I believe you can get rid of that line. Making things a little simpler.

Btw. is [sponsoring](https://github.com/sponsors/brillout) something your company would be up to? I'd then gladly implement #1056 sooner rather than later (I think it's doable within a week).

**Reply by openscript** [reply]
Thanks for your answer. We do sponsor projects like yours and charity. I'm currently evaluating it in one of my pet projects, but as soon as we earn money with vite-plugin-ssr we are happy to share the cake :)

**Reply by brillout** [reply]
Sounds good :)

Keep me updated how using vite-plugin-ssr (doesn't) work out for you. I'm happy to make adjustments. Also let me know if you believe #1056 is needed.

(Btw. I'm sometimes in Zurich as my borther lives there.)

**Reply by openscript** [reply]
More than happy to cowork in our office or have a beer if you around in Zurich!

**Reply by brillout** [reply]
I'll let you know if I'm around! (I don't have Zurich plans this year yet but something may come up.)

**Comment by samuelstroschein** [comment]
What a coincidence :D

1. @openscript (besides inlang codebase) what made you evaluate vite-plugin-ssr?
2. I hacked markdown outside of root support for the inlang repo with this file https://github.com/inlang/inlang/blob/main/documentation/tableOfContents.ts. Hacky but works

**Reply by openscript** [reply]
Hello @samuelstroschein 😄

1. I'm looking for an alternative to Gatsby. I like Gatsby's data layer, but we all know the rest of the story. I think it's easy to build something like Gatsby's data layer with vite-plugin-ssr.
2. Of course I glimpsed how you used vite-plugin-ssr in inlang, but I need something more flexible. Maybe I can soon upstream a bit to inlang as well.

**Reply by samuelstroschein** [reply]
> I'm looking for an alternative to Gatsby.  

What alternatives are you considering?

**Reply by openscript** [reply]
Well, the obvious candidates like Next.js, Remix and Astro. From these three I liked Astro the most, but some organizations have their own component library and with some sort of CSS-in-JS, what doesn't seem to be supported by Astro right now. Also I estimate the migration effort to VPS less, than to Astro. I like the Astro Island and the overall type safety.

**Reply by samuelstroschein** [reply]
Interesting. I recently looked at Astro again and "don't get it". 

- custom DSL (astro files) that increases learning and maintenance burden with no clear upside 
- limited to static sites and yet not limited (weird mix of "what is astro helpful for")

**Reply by openscript** [reply]
Yes the custom DSL would lead to a bigger migration effort and in general would like to not have another syntax in the mix.
